modified:   pycosep/community_separability.py

ywu57232 opened the pull request:
1) changed the computing of projection from using a loop to using matrix multiplication.
    2) changed conditional setting in aupr evaluation to equal the Matlab code.
    3) changed the way to generate random permutation indices to match the Matlab code.
    4) add a parameter to run concorde on different machine platforms.

ywu57232 committed:
modified:   pycosep/community_separability.py

    1) changed the computing of projection from using a loop to using matrix multiplication.
    2) changed conditional setting in aupr evaluation to equal the Matlab code.
    3) changed the way to generate random permutation indices to match the Matlab code.
    4) add a parameter to run concorde on different machine platforms.

aacevedot commented on pycosep/community_separability.py:
```diff
@@ -275,23 +278,22 @@ def _tsp_based_projection(pairwise_data, concorde_settings):
     # prepare TSP file
     with open(file_tsp_path, 'w') as file:
-        file.write("NAME : TSPS Concorde\n")
+        file.write(f"NAME : TSPS Concorde\n")
```
Do you have the latest `main` in your branch? We do not need the formatting in cases where there is no placeholder (same for the other changes below).

ywu57232 replied:
This was caused by editing the code from an outdated main. I have changed to the latest one.

aacevedot commented on pycosep/community_separability.py:
```diff
@@ -275,23 +278,22 @@ def _tsp_based_projection(pairwise_data, concorde_settings):
 
-    result = subprocess.run(command, cwd=concorde_settings.temp_path, capture_output=True, text=True, check=False,
-                            shell=True)
+    result = subprocess.run(command, cwd=concorde_settings.temp_path, capture_output=True, text=True, check=False, shell=(sys.platform=='linux' or sys.platform=='darwin' or sys.platform=='cygwin' )) # run corconde on different platforms
```
Please use the way I did on my `main` branch; a simple `True` does the job.

ywu57232 replied:
I changed to what you suggested.

aacevedot commented on pycosep/community_separability.py:
```diff
@@ -327,8 +329,8 @@ def _randomize_communities(communities, total_permutations):
     for ix in range(total_permutations):
-        np.random.seed(ix)
-        positions = np.random.permutation(total_communities)
+        np.random.seed(ix+1)            # to equal the Matlab random permutation indices
```
Do we want to match MATLAB's seeding for a better debugging experience, or does it impact the results? 🤔

ywu57232 replied:
Partly for the debugging purpose. I think it is fine to use the original random,  since the fluctuation of the results is in an acceptable range compared to Matlab results.

aacevedot commented on pycosep/community_separability.py:
```diff
@@ -541,9 +543,8 @@ def compute_separability(embedding, communities, positives=None, variant=Separab
-                separability_path = _convert_tour_to_one_dimension(best_tour, pairwise_data, pairwise_communities)
-                scores, source_nodes, target_nodes, edge_weights, cut_start_node, cut_end_node = separability_path
-
+                scores, source_nodes, target_nodes, edge_weights, cut_start_node, cut_end_node = _convert_tour_to_one_dimension(
```
Hmm, this and others are changes from the previous version. Please merge the latest `main` into yours so you do not have to carry this over.

ywu57232 replied:
Yes. This was caused by editing the code from an outdated main. I have changed to the latest one.

aacevedot reviewed (changes requested):
How about the unit tests? Did you check and update the values in there? I can see that the ones without permutations are passing, but the ones with permutations are not. Please, could you take a look? I enabled the CI for this PR, so eventually, the tests should run when you push new changes.

ywu57232 committed:
modified:   pycosep/community_separability.py

    1) changed the computing of projection from using a loop to using matrix multiplication.
    2) changed conditional setting in aupr evaluation to equal the Matlab code.
    3) changed the way to generate random permutation indices to match the Matlab code.
    4) add a parameter to run concorde on different machine platforms.

ywu57232 committed:
modified:   pycosep/community_separability.py

    1) changed the computing of projection from using a loop to using matrix multiplication.
    2) changed conditional setting in aupr evaluation to equal the Matlab code.
    3) changed the way to generate random permutation indices to match the Matlab code.

ywu57232 committed:
modified:   tests/test_cps_community_separability.py

ywu57232 committed:
Changes to be committed:
	modified:   tests/test_cps_community_separability.py
	modified:   tests/test_ldps_community_separability.py
	modified:   tests/test_tsps_community_separability.py

    These are changed to pass the unit tests.

aacevedot commented on pycosep/community_separability.py:
```diff
@@ -327,8 +329,8 @@ def _randomize_communities(communities, total_permutations):
-        np.random.seed(ix)
-        positions = np.random.permutation(total_communities)
+        np.random.seed(ix+1)            # to equal the Matlab random permutation indices
+        positions = np.argsort(np.random.random((1,total_communities)))[0]
```
Why do you need this? 🤔

aacevedot replied:
What is the benefit of using this approach versus `np.random.permutation`?

aacevedot replied:
Ahh I think I got it. The Fisher-Yates shuffle in NumPy does not explicitly generate and sort a list of random numbers as MATLAB does, and your approach seems to do it the same way. However, I am a bit concerned about performance.

aacevedot commented:
Since the CI in this PR is failing, it cannot be merged. However, I applied the changes to #9 and fixed the failing tests in there. Therefore, I will close this PR. Thanks, @ywu57232, for the improvements!
